File: Chapter 02/foo_class.cpp
#include<iostream>
class Foo{
    private: 
        // members
        int bar;
        int baz;
    // constructor
    public:
    Foo(){
        bar = 45;
        baz = 32;
    }
    // bounded method
    void do_something(){
        bar = baz + bar;
        baz--;
    }
    // destructor
    ~Foo(){
        // integers don't need to be freed 
        std::cout << "Bye " << bar << " and " << baz << std::endl;
    }

};

int main(){
    Foo f = Foo();
    f.do_something();
}

File: Chapter 02/foo_class.rs
struct Foo{
    // members => fields
    bar: i32, 
    baz: i32, 
}

// bounded methods => implementations 
// to declare use keyword impl
impl Foo{
    // to create instance return Self 
    // to use instance use self, &self or &mut self in the parameters 

    // constructor 
    pub fn new()-> Self{
        Foo{
            bar: 45, 
            baz: 32
        }
    }
    // bounded method
    pub fn do_something(&mut self){
        self.bar = self.bar + self.baz;
        self.baz -= 1;
    }
}

// structs will automatically be deallocated 
// but we can still create a destructor by
// implementing Drop for Foo
impl Drop for Foo{
    fn drop(&mut self){
        println!("Bye {} and {}", self.bar, self.baz)
    }
}

fn main(){
    let mut foo = Foo::new();
    foo.do_something();
}